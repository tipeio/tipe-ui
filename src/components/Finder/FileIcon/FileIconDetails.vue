<template>
  <div
    :data-tipe-ui="$options.name"
    :class="rootClassObject"
    @mouseover="hovered = true"
    @mouseout="hovered = false"
    @click="$emit('click', $event)"
  >
    <button
      class="details-button"
      @click="$emit('click')"
    >
      <div class="details-layout">
        <div class="icon-container">
          <tipe-icon :icon="icon"/>
        </div>
        <div class="label">
          {{ label }}
        </div>
        <div class="created-by">
          Created by {{ author }}
        </div>
        <div class="dot">&#8226;</div>
        <div class="updated-at">
          Edited {{ updatedAt }}
        </div>
      </div>
    </button>
    <button
      v-if="hovered && !picker"
      class="trash"
      @click="$emit('delete')"
    >
      <tipe-icon icon="trash"/>
    </button>
    <button
      v-if="hovered && !picker"
      class="edit"
      @click="$emit('edit')"
    >
      <tipe-icon icon="edit"/>
    </button>
  </div>
</template>

<script>
import vueTypes from 'vue-types'
import TipeIcon from '@/components/Icon'
import interfaces from '@tipe/tipe-interfaces'

export default {
  name: 'TipeFileIconDetails',
  components: {
    TipeIcon
  },
  props: {
    picker: vueTypes.bool.def(false),
    selected: vueTypes.bool.def(false),
    ...interfaces.file
  },
  data() {
    return {
      hovered: false
    }
  },
  computed: {
    author() {
      return `${this.createdBy.firstName} ${this.createdBy.lastName}`
    },
    rootClassObject: props => ({
      selected: props.selected
    })
  }
}
</script>

<style lang="postcss" scoped>
[data-tipe-ui='TipeFileIconDetails'] {
  display: grid;
  grid-template-columns: 1fr 3rem 3rem;
  grid-template-rows: 4.8125rem;
  align-items: center;
  justify-content: flex-start;
  border-bottom: 1px solid var(--gray-light);
}

[data-tipe-ui='TipeFileIconDetails']:first-child {
  border-top: 1px solid var(--gray-light);
}

[data-tipe-ui='TipeFileIconDetails']:nth-child(odd) {
  background-color: rgba(0, 0, 0, 0.01);
}

[data-tipe-ui='TipeFileIconDetails']:hover {
  background-color: rgba(255, 255, 255, 0.9);
}

[data-tipe-ui='TipeFileIconDetails'].selected {
  background-color: rgba(0, 0, 0, 0.06);
}

.details-layout {
  display: grid;
  grid-template-areas: 'icon label label label' 'icon created-by dot updated-at';
  grid-template-columns: 3rem auto auto auto auto;
  grid-template-rows: 1fr 1fr;
  overflow: hidden;
  justify-content: flex-start;
}

.icon-container {
  grid-area: icon;
  height: 100%;
  width: 100%;
  display: flex;
  align-items: center;
  justify-content: center;
}

.label {
  grid-area: label;
  grid-column: 1fr;
  display: grid;
  white-space: nowrap;
  font-size: 0.875rem;
  white-space: nowrap;
  letter-spacing: 0.008125rem;
  color: var(--text-gray);
  font-weight: 500;
  align-items: center;
  justify-content: flex-start;
}

.created-by {
  grid-area: created-by;
  grid-column: auto;
  display: grid;
  align-items: center;
  justify-content: flex-start;
  font-size: 0.75rem;
  white-space: nowrap;
  letter-spacing: 0.006875rem;
  color: var(--text-gray);
}

.updated-at {
  grid-area: updated-at;
  grid-column: 1fr;
  display: grid;
  align-items: center;
  justify-content: flex-start;
  font-size: 0.75rem;
  white-space: nowrap;
  letter-spacing: 0.006875rem;
  color: var(--text-gray);
}

.dot {
  grid-area: dot;
  color: var(--text-gray);
  font-size: 0.8rem;
  padding: 0 0.4rem;
}

button {
  margin: 0;
  padding: 0;
  border: none;
  outline: none;
  background-color: transparent;
  height: 100%;
  width: 100%;
}
</style>
