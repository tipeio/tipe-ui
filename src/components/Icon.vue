<template>
  <div
    :class="size"
    :data-tipe-ui="$options.name"
  />
</template>

<script>
import vueTypes from 'vue-types'

const sizes = ['small', 'medium', 'large']

export default {
  name: 'TipeIcon',
  props: {
    size: vueTypes.oneOf(sizes).def('medium')
  }
}
</script>

<style scoped>
[data-tipe-ui='TipeIcon'] {
  background-color: red;
}

[data-tipe-ui='TipeIcon'].small {
  height: 1rem;
  width: 1rem;
}

[data-tipe-ui='TipeIcon'].medium {
  height: 2rem;
  width: 2rem;
}

[data-tipe-ui='TipeIcon'].large {
  height: 3rem;
  width: 3rem;
}
</style>
