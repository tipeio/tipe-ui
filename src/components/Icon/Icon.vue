<template>
  <object
    :data="src"
    type="image/svg+xml">
    <img :src="src" >
  </object>
</template>

<script>
import vueTypes from 'vue-types'

export default {
  name: 'TipeIcon',
  props: {
    icon: vueTypes.string
  },
  computed: {
    src() {
      return require(`../../icons/${this.icon}.svg`)
    }
  }
}
</script>
