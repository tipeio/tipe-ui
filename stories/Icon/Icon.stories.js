import { storiesOf } from '@storybook/vue'

import TipeIcon from '../../src/components/Icon'

const style = () => ({
  height: '1rem',
  width: '1rem'
})

const icons = [
  'check',
  'document',
  'folder',
  'key',
  'number',
  'share',
  'switch',
  'user',
  'color',
  'edit',
  'gear',
  'link',
  'pin',
  'sign-out',
  'text',
  'view',
  'copy',
  'flow',
  'grid',
  'list',
  'search',
  'star-active',
  'trash',
  'wrench',
  'credit-card',
  'folder-open',
  'group',
  'markdown',
  'media',
  'send',
  'star',
  'user-circle',
  'options',
  'plus',
  'cross',
  'arrow-down',
  'arrow-up',
  'arrow-right',
  'arrow-left'
]

storiesOf('Icon', module).add('icons', () => ({
  components: { TipeIcon },
  computed: { style },
  template:
    '<div>' +
    icons.reduce((template, icon) => {
      return (
        template +
        `<tipe-icon :style="style" icon="${icon}" color="rgb(255,0,0)" />`
      )
    }, '') +
    '</div>'
}))
