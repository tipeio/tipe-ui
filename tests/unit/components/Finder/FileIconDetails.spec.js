import { shallowMount } from '@vue/test-utils'
import { createRenderer } from 'vue-server-renderer'
import moment from 'moment'
import FileIconDetails from '@/components/Finder/FileIcon/FileIconDetails'
import Icon from '@/components/Icon'
import fileMock from '@/mocks/File'

describe('FileIconDetails', () => {
  describe('<template>', () => {
    it('matches previous snapshot', () => {
      const renderer = createRenderer()
      const propsData = {
        file: {
          label: 'hello world',
          type: 'document',
          updatedAt: Date.now(),
          createdBy: {
            firstName: 'test',
            lastName: 'test'
          }
        }
      }
      const wrapper = shallowMount(FileIconDetails, { propsData })

      renderer.renderToString(wrapper.vm, (err, str) => {
        if (err) throw new Error(err)
        expect(str).toMatchSnapshot()
      })
    })

    it('has correct data-tipe-iu attibute', () => {
      const propsData = { file: fileMock() }
      const wrapper = shallowMount(FileIconDetails, { propsData })

      expect(wrapper.attributes()['data-tipe-ui']).toBe('TipeFileIconDetails')
    })
  })

  describe(':props', () => {
    it(':file - should render label prop as label', () => {
      const propsData = { file: fileMock() }
      const wrapper = shallowMount(FileIconDetails, { propsData })

      expect(wrapper.find('.label').text()).toEqual(propsData.file.label)
    })

    it(':file.type(folder) - should pass "folder" to Icon component\'s :icon prop', () => {
      const propsData = { file: fileMock({ type: 'folder' }) }
      const wrapper = shallowMount(FileIconDetails, { propsData })

      expect(wrapper.find(Icon).props().icon).toEqual('folder')
    })

    it(':file.type(document) - should pass "file" to Icon component\'s :icon prop', () => {
      const propsData = { file: fileMock({ type: 'document' }) }
      const wrapper = shallowMount(FileIconDetails, { propsData })

      expect(wrapper.find(Icon).props().icon).toEqual('file')
    })

    it(':file.updatedAt - should display last updated delta', () => {
      const propsData = { file: fileMock() }
      const wrapper = shallowMount(FileIconDetails, { propsData })

      expect(wrapper.find('.updated-at').text()).toMatch(
        new RegExp(moment(propsData.file.updatedAt).fromNow())
      )
    })

    it(":file.createdBy - should display author's name", () => {
      const propsData = { file: fileMock() }
      const wrapper = shallowMount(FileIconDetails, { propsData })

      expect(wrapper.find('.created-by').text()).toMatch(
        new RegExp(
          `${propsData.file.createdBy.firstName} ${
            propsData.file.createdBy.lastName
          }`
        )
      )
    })
  })
})

describe('@events', () => {
  it('@hovered - should show trash button', () => {
    const propsData = { file: fileMock() }
    const wrapper = shallowMount(FileIconDetails, { propsData })

    expect(wrapper.contains('.trash')).toEqual(false)
    wrapper.trigger('mouseover')
    expect(wrapper.contains('.trash')).toEqual(true)
  })

  it('@hovered - should show edit button', () => {
    const propsData = { file: fileMock() }
    const wrapper = shallowMount(FileIconDetails, { propsData })

    expect(wrapper.contains('.edit')).toEqual(false)
    wrapper.trigger('mouseover')
    expect(wrapper.contains('.edit')).toEqual(true)
  })

  it('@click - should emit click event', () => {
    const propsData = { file: fileMock() }
    const wrapper = shallowMount(FileIconDetails, { propsData })

    wrapper.find('.details-button').trigger('click')
    expect(wrapper.emitted('click')).toBeTruthy()
  })

  it('@delete - should emit delete event', () => {
    const propsData = { file: fileMock() }
    const wrapper = shallowMount(FileIconDetails, { propsData })

    wrapper.trigger('mouseover')
    wrapper.find('.trash').trigger('click')
    expect(wrapper.emitted('delete')).toBeTruthy()
  })

  it('@edit - should emit delete event', () => {
    const propsData = { file: fileMock() }
    const wrapper = shallowMount(FileIconDetails, { propsData })

    wrapper.trigger('mouseover')
    wrapper.find('.edit').trigger('click')
    expect(wrapper.emitted('edit')).toBeTruthy()
  })
})
