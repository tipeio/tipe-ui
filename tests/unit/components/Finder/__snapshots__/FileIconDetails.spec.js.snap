// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileIconDetails <template> matches previous snapshot 1`] = `
<div data-tipe-ui="TipeFileIconDetails">
  <button class="details-button">
    <div class="details-layout">
      <div class="icon-container">
        <!---->
      </div>
      <div class="label">
        hello world
      </div>
      <div class="created-by">
        Created by test test
      </div>
      <div class="dot">
        <!---->
      </div>
      <div class="updated-at">
        Edited a few seconds ago
      </div>
    </div>
  </button>
  <!---->
  <!---->
</div>
`;
